sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -309,53 +306,112 @@ private IndexItemResponse indexItem(ShardUpsertRequest request,
     }
 
     @VisibleForTesting
-    protected IndexItemResponse indexItem(ShardUpsertRequest request,
-                                          ShardUpsertRequest.Item item,
-                                          IndexShard indexShard,
-                                          boolean tryInsertFirst,
-                                          UpdateSourceGen updateSourceGen,
-                                          InsertSourceGen insertSourceGen,
-                                          @Nullable ReturnValueGen returnGen,
-                                          boolean isRetry) throws Exception {
-        final long seqNo;
-        final long primaryTerm;
-        final long version;
-        Doc updatedDoc = null;
-        if (tryInsertFirst) {
-            version = request.duplicateKeyAction() == DuplicateKeyAction.OVERWRITE
-                ? Versions.MATCH_ANY
-                : Versions.MATCH_DELETED;
-            seqNo = SequenceNumbers.UNASSIGNED_SEQ_NO;
-            primaryTerm = SequenceNumbers.UNASSIGNED_PRIMARY_TERM;
-            try {
-                item.source(insertSourceGen.generateSourceAndCheckConstraints(item.insertValues()));
-            } catch (IOException e) {
-                throw ExceptionsHelper.convertToElastic(e);
+    protected IndexItemResponse insert(ShardUpsertRequest request,
+                                       ShardUpsertRequest.Item item,
+                                       IndexShard indexShard,
+                                       boolean isRetry,
+                                       @Nullable ReturnValueGen returnGen,
+                                       @Nullable InsertSourceGen insertSourceGen) throws Exception {
```

Comment:
Please remove `Nullable` annotation as it is expected to not be null by next assertion

Reply:
You are right, I had a different understanding of what `Nullable` means.